crates/bindings-csharp/Runtime/Internal/ITable.cs

```diff
@@ -22,31 +22,45 @@ public class Enumerator(FFI.RowIter handle) : IDisposable
 
             public bool MoveNext()
             {
+                if (handle.Equals(FFI.RowIter.INVALID))
+                {
+                    return false;
+                }
+
                 uint buffer_len;
                 while (true)
                 {
                     buffer_len = (uint)buffer.Length;
-                    try
+                    var ret = FFI._row_iter_bsatn_advance(handle, buffer, ref buffer_len);
+                    if (ret <= 0)
                     {
-                        FFI._iter_advance(handle, buffer, ref buffer_len);
+                        Current = new byte[buffer_len];
+                        Array.Copy(buffer, 0, Current, 0, buffer_len);
                     }
-                    catch (BufferTooSmallException)
+                    switch (ret)
                     {
-                        buffer = new byte[buffer_len];
-                        continue;
+                        // Iterator exhausted, we're done.
+                        case -1:
+                            handle = FFI.RowIter.INVALID;
+                            return buffer_len != 0;
+                        case 0:
+                            return buffer_len != 0;
+                        case (short)(ushort)FFI.Errno.NO_SUCH_ITER:
+                            throw new NoSuchIterException();
+                        case (short)(ushort)FFI.Errno.BUFFER_TOO_SMALL:
+                            buffer = new byte[buffer_len];
```

Comment:
Shouldn't this copy the original buffer? Not sure if any data is written to the buffer at that point or not.